### PR TITLE
Retrained L1 Tau Trigger Integration

### PR DESCRIPTION
**Pull request by Duchstf**
#### PR description:

Hello, this PR updates the Tau NN training in the L1 trigger. The main changes include:

- Adding HW option for the tau as well as the new e/m seeding for tau jets.
- Updated Tau NN training with the lastest training samples.
- GT formatting for Tau Jets.

#### PR validation:

The improvements in performance is documented in the following talks:

- https://indico.cern.ch/event/1246470/
- https://indico.cern.ch/event/1271855/#202-performance-report-algorit

The new cmssw performance has been cross-checked by the L1 team and the results are included in the [2023 L1 Annual Review.](https://indico.cern.ch/event/1257943/contributions/5343040/attachments/2633523/4562436/AlgorithmsAndPerformance_2023_2.pdf) 

The original PR with related discussions/checks is here: https://github.com/cms-l1t-offline/cmssw/pull/1087

### CODE REVIEW

**Comment by aloeliger** [comment]
@Duchstf I have been searching both the L1 fork and CMSSW and cannot find any instance of a `TauNNTFCache` does this code exist in the integration branch somewhere that I have missed? Is this something that you have introduced?

**Comment by qvyz** [comment]
> @Duchstf I have been searching both the L1 fork and CMSSW and cannot find any instance of a `TauNNTFCache` does this code exist in the integration branch somewhere that I have missed? Is this something that you have introduced?

Hi, I was running in this issue when trying to test the GT emulator:
The definition of TauNNTFCache is present in the snapshot branch 3  in the l1t offline repo.
(https://github.com/cms-l1t-offline/cmssw/blob/9fc02faa5eff9a7ad19bb9cd43e6c0cfb9f78867/L1Trigger/Phase2L1ParticleFlow/interface/TauNNId.h#L8-L11) maybe this file was forgotten

**Comment by aloeliger** [comment]
Okay, this particular structure was actual removed, via agreement about 5 months ago in https://github.com/cms-sw/cmssw/pull/40333. It seems like this change was never backported, and this branch was never properly rebased onto a more recent branch with these changes, and adapted for them. 

@Duchstf @EmyrClement  You need to rebase these changes on to a more recent branch, and you will very likely need to fix some merge conflicts, retest the compilation of the emulator, and adapt it to the changes that have happened in the main fork.

something like:
```
cmsrel CMSSW_13_2_0_pre1
cd CMSSW_13_2_0_pre1/src/
cmsenv && git cms-init
git cms-rebase -u 
Duchstf:Tau_L1_2023_cmssw
```

at which point you will likely need to introduce some fixes, and make sure it all compiles again. **Please make sure this work is backed up in a backup branch in case anything goes wrong!**. Once it is properly adjusted, try pushing the changes back here.

Just one of the perils of trying to maintain a parallel git fork/branch I guess.

**Comment by Duchstf** [comment]
@aloeliger Thank you so much for tracking down the error!!

**Comment by epalencia** [comment]
Please test

**Comment by EmyrClement** [comment]
@Duchstf can you confirm if [this change](https://github.com/cms-l1t-offline/cmssw/pull/1087/commits/f30191306006dd9151ce4de6b3b97e2c7e25393d) is already included here or not?  We should make sure it is before proceeding.

**Comment by Duchstf** [comment]
@EmyrClement I added it now! Sorry I didn't get the chance to make the change before the tests were ran :(

**Comment by EmyrClement** [comment]
Just to note that the `.pb` files used when running the simulation workflow (which is no longer the default) are updated by [this PR to cms-data](https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/7).

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 1]
```diff
@@ -149,9 +186,163 @@ void L1NNTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
...
+  L1TauEmu::pt_t mass = l1ct::pt_t(tmpmass1);
+  if (pt < fSeedPt_)
+    return;
+  result_t NN = fTauNNIdHW_->compute(seed, parts);
+  input_t* lNNVector = fTauNNIdHW_->NNVectorVar();
+  float pNNVec[80];
+  for (unsigned i0 = 0; i0 < 80; i0++)
+    pNNVec[i0] = float(lNNVector[i0]);
+  L1TauEmu::etaphi_t eta = etaphi_t(seed.eta() * L1TauEmu::etaphi_base);
+  L1TauEmu::etaphi_t phi = etaphi_t(seed.phi() * L1TauEmu::etaphi_base);
```
Comment (perrotta) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:264]:
```suggestion
```
These two aren't needed any more now, or am I wrong?

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc [hunk 2]
```diff
@@ -149,9 +186,163 @@ void L1NNTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
...
+      if (parts[i0].id() == l1t::PFCandidate::Electron || parts[i0].id() == l1t::PFCandidate::ChargedHadron ||
+          parts[i0].id() == l1t::PFCandidate::Photon) {
+        lId++;
+        pt = pt + L1TauEmu::pt_t(parts[i0].pt());
+        float deta = parts[i0].eta() - eta_1;
+        float dphi = parts[i0].phi() - phi_1;
+        float dr2 = deta * deta + dphi * dphi;
+        pz_tot = pz_tot + (parts[i0].pt()) * (1 - dr2 * 0.5);
+        py_tot = py_tot + (parts[i0].pt()) * dphi;  //sin(dphi ));
+        px_tot = px_tot + (parts[i0].pt()) * deta;  //sin(deta ));
```
Comment (perrotta) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:233]:
```suggestion
```
also these do not seem to be used anywhere...
After a very quick glance this code still need quite some cleaning

**Comment by Duchstf** [comment]
@perrotta Thanks for the quick looks!! I removed more codes that are unnecessary, let me know if you have more comments!! 

Duc.

**Comment by perrotta** [comment]
> @perrotta Thanks for the quick looks!! I removed more codes that are unnecessary, let me know if you have more comments!!
> 
> Duc.

Thank you Duc.
Not now, I'd let the usual reviewer to comment it first, release managers enter later in the game normally. I only went through it by chance and I noticed that some cleaning should have been done.

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
@Duchstf: The compiler finds unused variables:

```
L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_2_X_2023-06-06-2300/src/L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:209:9: warning: unused variable 'eta_1' [-Wunused-variable]
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_2_X_2023-06-06-2300/src/L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc:210:9: warning: unused variable 'phi_1' [-Wunused-variable]
```

These need to be removed.

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
This PR is (somewhat) dependent on https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/7

**Comment by aloeliger** [comment]
Please test with https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/7

**Comment by aloeliger** [comment]
+l1

This PR has been reviewed by L1 as a part of https://github.com/cms-l1t-offline/cmssw/pull/1087, and is necessary for the ongoing GT emulator PR https://github.com/cms-sw/cmssw/pull/41808 . At this point I think we would like to hear from others.

**Comment by aloeliger** [comment]
@Duchstf Have you had a chance to look at upgrade's review comments?

**Comment by Duchstf** [comment]
> @Duchstf Have you had a chance to look at upgrade's review comments?

Hi @aloeliger, I'm looking at it now. However I'm wondering if you could tell me what's the correct recipe to set up my branch with the main CMSSW is, I was using your suggestions earlier:

```
cmsrel CMSSW_13_2_0_pre1
cd CMSSW_13_2_0_pre1/src/
cmsenv && git cms-init
git cms-rebase-topic-u Duchstf:Tau_L1_2023_cmssw
```

However it gives me this errors:

```
(base) bash-4.2$ git cms-rebase-topic -u Duchstf:Tau_L1_2023_cmssw
From .
 * [new ref]                            -> rebase-attempt
From https://github.com/Duchstf/cmssw
 * [new branch]              Tau_L1_2023_cmssw -> Duchstf/Tau_L1_2023_cmssw
Switched to branch 'rebase-attempt'
Switched to a new branch 'Tau_L1_2023_cmssw'
Created branch Tau_L1_2023_cmssw to follow Tau_L1_2023_cmssw from repository Duchstf
Auto-merging DataFormats/L1TParticleFlow/interface/gt_datatypes.h
Auto-merging DataFormats/L1TParticleFlow/src/classes_def.xml
CONFLICT (content): Merge conflict in DataFormats/L1TParticleFlow/src/classes_def.xml
error: could not apply bd46e3501b7... Merge L1 Tau Trigger changes
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply bd46e3501b7... Merge L1 Tau Trigger changes
```

I'm pretty confused because the codes passed all the checks, so I don't really know where the problem is.

**Comment by AdrianoDee** [comment]
Hi @Duchstf this is probably due to the fact that you are using `CMSSW_13_2_0_pre1` that is not that new. I'd suggest you to use one of the latest IBs to be better aligned to the latest developments (such as `MSSW_13_2_X_2023-06-20-1100`). You can always check which is the latest with something like `scram list | grep 13_2_X`.

**Comment by aloeliger** [comment]
@Duchstf It is probably simpler to just check out the branch you already have, instead of rebasing it again?

```
cmsrel CMSSW_13_2_0_pre1
cd CMSSW_13_2_0_pre1/src/
cmsenv && git cms-init
git cms-checkout-topic-u Duchstf:Tau_L1_2023_cmssw
```

I tend to forget about the `cms-checkout-topic` command, but it is what you're looking for here I think.

**Comment by Duchstf** [comment]
@AdrianoDee I think I addressed all the comments you have, let me know if I need anything else!!

**Comment by aloeliger** [comment]
please test

**Comment by srimanob** [comment]
@cmsbuild please test

It seems to be a glitch, let's try again.

**Comment by srimanob** [comment]
CMSSW_13_2_X_2023-06-20-2300 is broken, we should wait until the IB comes.

**Comment by aloeliger** [comment]
> CMSSW_13_2_X_2023-06-20-2300 is broken, we should wait until the IB comes.

I see [CMSSW_13_2_X_2023-06-21-1100](https://github.com/cms-sw/cmssw/releases/tag/CMSSW_13_2_X_2023-06-21-1100) and [CMSSW_13_2_X_2023-06-21-1600](https://github.com/cms-sw/cmssw/releases/tag/CMSSW_13_2_X_2023-06-21-1600) in the tags. Out of curiosity, when does buildbot pick up a new tag for IB's?

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by AdrianoDee** [comment]
+upgrade

**Comment by aloeliger** [comment]
+l1

**Comment by rappoccio** [comment]
Can someone explain why hundreds of magic numbers are added to a dataformats class? This looks extremely suboptimal. At the very least it should be configurable, but in the best case scenario this should come from the CondDB.

**Comment by srimanob** [comment]
I would propose to open git issue to follow up on performance and code cleaning. I agree that some part can migrate to db or cms data. However, this PR does not introduce new data file, but updates the one which are in the release already.

**Comment by aloeliger** [comment]
@rappoccio What you are probably referring to are the neural network weights for the phase 2 taus? This phase 2 project was started before L1 settled on a definite HLS4ML emulation strategy developed for CICADA and so I think just tried to insert the firmware model into CMSSW as-is, so updating a training relies on changing these weights this way.

@EmyrClement can comment norepinephrine, but I believe that it is on the to-do list to migrate emulation strategies eventually.

**Comment by rappoccio** [comment]
+1

**Pull request merged**